util/tracing/opt_trace.go

```diff
@@ -178,48 +184,18 @@ type CandidatePlanTrace struct {
 	MappingLogicalPlan string `json:"mapping"`
 }
 
-func newCandidatePlanTrace(trace *PlanTrace, logicalPlanKey string, bestKey map[string]struct{}) *CandidatePlanTrace {
-	selected := false
-	if _, ok := bestKey[CodecPlanName(trace.TP, trace.ID)]; ok {
-		selected = true
-	}
-	c := &CandidatePlanTrace{
-		MappingLogicalPlan: logicalPlanKey,
-	}
-	c.PlanTrace = trace
-	c.Selected = selected
-	for i, child := range c.Children {
-		if _, ok := bestKey[CodecPlanName(child.TP, child.ID)]; ok {
-			child.Selected = true
-		}
-		c.Children[i] = child
-	}
-	return c
-}
-
 // buildCandidatesInfo builds candidates info
 func (tracer *PhysicalOptimizeTracer) buildCandidatesInfo() {
-	if tracer == nil || len(tracer.State) < 1 {
+	if tracer == nil || len(tracer.Candidates) < 1 {
 		return
 	}
-	sCandidates := make([]*CandidatePlanTrace, 0)
-	dCandidates := make([]*CandidatePlanTrace, 0)
-	bestKeys := map[string]struct{}{}
-	for _, node := range tracer.Final {
-		bestKeys[CodecPlanName(node.TP, node.ID)] = struct{}{}
-	}
-	for logicalKey, pps := range tracer.State {
-		for _, pp := range pps {
-			c := newCandidatePlanTrace(pp, logicalKey, bestKeys)
-			if c.Selected {
-				sCandidates = append(sCandidates, c)
-			} else {
-				dCandidates = append(dCandidates, c)
+	for _, candidate := range tracer.Candidates {
+		for _, plan := range tracer.Final {
```

Comment:
loop-in-loop can be replaced with two loop and check an hashset to find the exist.